### PR TITLE
fix for when expressions issues

### LINKED ISSUE #3196: WhenExpressions are neglected when finally tasks are present
# Expected Behavior
WhenExpressions should work when finally tasks are present

# Actual Behavior
WhenExpressions will not work when finally tasks are present

# Steps to Reproduce the Problem

1. Create a Pipeline with tasks, when expressions and finally tasks 
2. Start the pipeline

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
  Client Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0+b3b92b2", GitCommit:"b3b92b2", GitTreeState:"clean", BuildDate:"2020-07-15T09:27:21Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.3+8b0a82f", GitCommit:"8b0a82f", GitTreeState:"clean", BuildDate:"2020-07-10T05:34:00Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version:
  ```
  $ tkn version
  Client version: 0.12.0
  Pipeline version: v0.16.0
  Triggers version: unknown
  ```

**Command Execution Logs**:
Execution logs contain the output of both `pipeline with finally task` and `pipeline without finally task`
```
[varadhya@localhost workspace-testing]$ tkn t ls
NAME          DESCRIPTION          AGE
final-task    This is first task   8 minutes ago
first-task    This is first task   17 minutes ago
second-task   This is first task   17 minutes ago
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn p ls
No Pipelines found
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ cat pipeline.yaml 
apiVersion: tekton.dev/v1beta1 
kind: Pipeline 
metadata:
  name: test-when-expression-pipeline
spec:
  params:
    - name: status
  tasks:
    - name: first-task
      taskRef:
        name: first-task
        kind: Task
    - name: second-task
      taskRef:
        name: second-task
        kind: Task
      when:
        - input: "$(params.status)"
          operator: in
          values: ["Success", "success"]
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ kubectl apply -f pipeline.yaml 
pipeline.tekton.dev/test-when-expression-pipeline created
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn p start test-when-expression-pipeline 
? Value for param `status` of type `string`? Failure
PipelineRun started: test-when-expression-pipeline-run-pvlw9

In order to track the PipelineRun progress run:
tkn pipelinerun logs test-when-expression-pipeline-run-pvlw9 -f -n veeresh-testing
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn pr describe test-when-expression-pipeline-run-pvlw9 -o yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1beta1","kind":"Pipeline","metadata":{"annotations":{},"name":"test-when-expression-pipeline","namespace":"veeresh-testing"},"spec":{"params":[{"name":"status"}],"tasks":[{"name":"first-task","taskRef":{"kind":"Task","name":"first-task"}},{"name":"second-task","taskRef":{"kind":"Task","name":"second-task"},"when":[{"input":"$(params.status)","operator":"in","values":["Success","success"]}]}]}}
  creationTimestamp: "2020-09-10T15:57:41Z"
  generateName: test-when-expression-pipeline-run-
  generation: 1
  labels:
    tekton.dev/pipeline: test-when-expression-pipeline
  managedFields:
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
      f:spec:
        .: {}
        f:params: {}
        f:pipelineRef:
          .: {}
          f:name: {}
    manager: tkn
    operation: Update
    time: "2020-09-10T15:57:41Z"
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          .: {}
          f:tekton.dev/pipeline: {}
      f:status:
        .: {}
        f:completionTime: {}
        f:conditions: {}
        f:pipelineSpec:
          .: {}
          f:params: {}
          f:tasks: {}
        f:skippedTasks: {}
        f:startTime: {}
        f:taskRuns:
          .: {}
          f:test-when-expression-pipeline-run-pvlw9-first-task-t4qk4:
            .: {}
            f:pipelineTaskName: {}
            f:status:
              .: {}
              f:completionTime: {}
              f:conditions: {}
              f:podName: {}
              f:startTime: {}
              f:steps: {}
              f:taskSpec:
                .: {}
                f:description: {}
                f:results: {}
                f:steps: {}
    manager: controller
    operation: Update
    time: "2020-09-10T15:57:53Z"
  name: test-when-expression-pipeline-run-pvlw9
  namespace: veeresh-testing
  resourceVersion: "354062"
  selfLink: /apis/tekton.dev/v1beta1/namespaces/veeresh-testing/pipelineruns/test-when-expression-pipeline-run-pvlw9
  uid: d94cb1a6-ac99-44f6-8389-eea2e5541d78
spec:
  params:
  - name: status
    value: Failure
  pipelineRef:
    name: test-when-expression-pipeline
  timeout: 1h0m0s
status:
  completionTime: "2020-09-10T15:57:53Z"
  conditions:
  - lastTransitionTime: "2020-09-10T15:57:53Z"
    message: 'Tasks Completed: 1 (Failed: 0, Cancelled 0), Skipped: 1'
    reason: Completed
    status: "True"
    type: Succeeded
  pipelineSpec:
    params:
    - name: status
      type: string
    tasks:
    - name: first-task
      taskRef:
        kind: Task
        name: first-task
    - name: second-task
      taskRef:
        kind: Task
        name: second-task
      when:
      - Input: $(params.status)
        Operator: in
        Values:
        - Success
        - success
  skippedTasks:
  - name: second-task
  startTime: "2020-09-10T15:57:41Z"
  taskRuns:
    test-when-expression-pipeline-run-pvlw9-first-task-t4qk4:
      pipelineTaskName: first-task
      status:
        completionTime: "2020-09-10T15:57:53Z"
        conditions:
        - lastTransitionTime: "2020-09-10T15:57:53Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: test-when-expression-pipeline-run-pvlw9-first-task-t4qk4--ngrw6
        startTime: "2020-09-10T15:57:41Z"
        steps:
        - container: step-shell-example
          imageID: docker.io/library/ubuntu@sha256:31dfb10d52ce76c5ca0aa19d10b3e6424b830729e32a89a7c6eee2cda2be67a5
          name: shell-example
          terminated:
            containerID: cri-o://5c0e1ab4325a6bfb5f3e31414e474a344e57cb1afd4f773e6b96a90fd124eb79
            exitCode: 0
            finishedAt: "2020-09-10T15:57:52Z"
            reason: Completed
            startedAt: "2020-09-10T15:57:52Z"
        taskSpec:
          description: This is first task
          results:
          - description: this stores the status
            name: status
          steps:
          - image: ubuntu
            name: shell-example
            resources: {}
            script: |
              #!/bin/bash
              echo "This is first task"
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ cat pipeline-finally.yaml 
apiVersion: tekton.dev/v1beta1 
kind: Pipeline 
metadata:
  name: test-when-expression-pipeline-finally
spec:
  params:
    - name: status
  tasks:
    - name: first-task
      taskRef:
        name: first-task
        kind: Task
    - name: second-task
      taskRef:
        name: second-task
        kind: Task
      when:
        - input: "$(params.status)"
          operator: in
          values: ["Success", "success"]
  finally:
    - name: finally-task
      taskRef:
        name: final-task
        kind: Task
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ kubectl apply -f pipeline-finally.yaml 
pipeline.tekton.dev/test-when-expression-pipeline-finally created
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn p start test-when-expression-pipeline-finally
? Value for param `status` of type `string`? Failure
PipelineRun started: test-when-expression-pipeline-finally-run-vgvsp

In order to track the PipelineRun progress run:
tkn pipelinerun logs test-when-expression-pipeline-finally-run-vgvsp -f -n veeresh-testing
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn pr ls
NAME                                              STARTED          DURATION     STATUS
test-when-expression-pipeline-finally-run-vgvsp   40 seconds ago   29 seconds   Succeeded
test-when-expression-pipeline-run-pvlw9           4 minutes ago    12 seconds   Succeeded(Completed)
[varadhya@localhost workspace-testing]$ 
[varadhya@localhost workspace-testing]$ tkn pr describe test-when-expression-pipeline-finally-run-vgvsp -o yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1beta1","kind":"Pipeline","metadata":{"annotations":{},"name":"test-when-expression-pipeline-finally","namespace":"veeresh-testing"},"spec":{"finally":[{"name":"finally-task","taskRef":{"kind":"Task","name":"final-task"}}],"params":[{"name":"status"}],"tasks":[{"name":"first-task","taskRef":{"kind":"Task","name":"first-task"}},{"name":"second-task","taskRef":{"kind":"Task","name":"second-task"},"when":[{"input":"$(params.status)","operator":"in","values":["Success","success"]}]}]}}
  creationTimestamp: "2020-09-10T16:01:02Z"
  generateName: test-when-expression-pipeline-finally-run-
  generation: 1
  labels:
    tekton.dev/pipeline: test-when-expression-pipeline-finally
  managedFields:
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
      f:spec:
        .: {}
        f:params: {}
        f:pipelineRef:
          .: {}
          f:name: {}
    manager: tkn
    operation: Update
    time: "2020-09-10T16:01:02Z"
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          .: {}
          f:tekton.dev/pipeline: {}
      f:status:
        .: {}
        f:completionTime: {}
        f:conditions: {}
        f:pipelineSpec:
          .: {}
          f:finally: {}
          f:params: {}
          f:tasks: {}
        f:startTime: {}
        f:taskRuns:
          .: {}
          f:test-when-expression-pipeline-finally-run-vgvsp-finally-t-6dwhq:
            .: {}
            f:pipelineTaskName: {}
            f:status:
              .: {}
              f:completionTime: {}
              f:conditions: {}
              f:podName: {}
              f:startTime: {}
              f:steps: {}
              f:taskSpec:
                .: {}
                f:description: {}
                f:results: {}
                f:steps: {}
          f:test-when-expression-pipeline-finally-run-vgvsp-first-tas-qbnrg:
            .: {}
            f:pipelineTaskName: {}
            f:status:
              .: {}
              f:completionTime: {}
              f:conditions: {}
              f:podName: {}
              f:startTime: {}
              f:steps: {}
              f:taskSpec:
                .: {}
                f:description: {}
                f:results: {}
                f:steps: {}
          f:test-when-expression-pipeline-finally-run-vgvsp-second-ta-8k64m:
            .: {}
            f:pipelineTaskName: {}
            f:status:
              .: {}
              f:completionTime: {}
              f:conditions: {}
              f:podName: {}
              f:startTime: {}
              f:steps: {}
              f:taskSpec:
                .: {}
                f:description: {}
                f:results: {}
                f:steps: {}
    manager: controller
    operation: Update
    time: "2020-09-10T16:01:31Z"
  name: test-when-expression-pipeline-finally-run-vgvsp
  namespace: veeresh-testing
  resourceVersion: "354942"
  selfLink: /apis/tekton.dev/v1beta1/namespaces/veeresh-testing/pipelineruns/test-when-expression-pipeline-finally-run-vgvsp
  uid: 5720518b-6dbc-46a4-80ff-6f262a9f6c7c
spec:
  params:
  - name: status
    value: Failure
  pipelineRef:
    name: test-when-expression-pipeline-finally
  timeout: 1h0m0s
status:
  completionTime: "2020-09-10T16:01:31Z"
  conditions:
  - lastTransitionTime: "2020-09-10T16:01:31Z"
    message: 'Tasks Completed: 3 (Failed: 0, Cancelled 0), Skipped: 0'
    reason: Succeeded
    status: "True"
    type: Succeeded
  pipelineSpec:
    finally:
    - name: finally-task
      taskRef:
        kind: Task
        name: final-task
    params:
    - name: status
      type: string
    tasks:
    - name: first-task
      taskRef:
        kind: Task
        name: first-task
    - name: second-task
      taskRef:
        kind: Task
        name: second-task
      when:
      - Input: $(params.status)
        Operator: in
        Values:
        - Success
        - success
  startTime: "2020-09-10T16:01:02Z"
  taskRuns:
    test-when-expression-pipeline-finally-run-vgvsp-finally-t-6dwhq:
      pipelineTaskName: finally-task
      status:
        completionTime: "2020-09-10T16:01:31Z"
        conditions:
        - lastTransitionTime: "2020-09-10T16:01:31Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: test-when-expression-pipeline-finally-run-vgvsp-finally-t-d6blv
        startTime: "2020-09-10T16:01:17Z"
        steps:
        - container: step-shell-example
          imageID: docker.io/library/ubuntu@sha256:31dfb10d52ce76c5ca0aa19d10b3e6424b830729e32a89a7c6eee2cda2be67a5
          name: shell-example
          terminated:
            containerID: cri-o://04027c0ed7ebed6e144e31224d66cf1b6de8ab0840b70ca0fcc1a8660896efa0
            exitCode: 0
            finishedAt: "2020-09-10T16:01:30Z"
            reason: Completed
            startedAt: "2020-09-10T16:01:30Z"
        taskSpec:
          description: This is first task
          results:
          - description: this stores the status
            name: status
          steps:
          - image: ubuntu
            name: shell-example
            resources: {}
            script: |
              #!/bin/bash
              echo "This is final task"
    test-when-expression-pipeline-finally-run-vgvsp-first-tas-qbnrg:
      pipelineTaskName: first-task
      status:
        completionTime: "2020-09-10T16:01:17Z"
        conditions:
        - lastTransitionTime: "2020-09-10T16:01:17Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: test-when-expression-pipeline-finally-run-vgvsp-first-tas-4xqgm
        startTime: "2020-09-10T16:01:02Z"
        steps:
        - container: step-shell-example
          imageID: docker.io/library/ubuntu@sha256:31dfb10d52ce76c5ca0aa19d10b3e6424b830729e32a89a7c6eee2cda2be67a5
          name: shell-example
          terminated:
            containerID: cri-o://15cb134c06d14d88ef7ad647a035fcefb0275f6114aee76e7887ee00dc735a90
            exitCode: 0
            finishedAt: "2020-09-10T16:01:17Z"
            reason: Completed
            startedAt: "2020-09-10T16:01:17Z"
        taskSpec:
          description: This is first task
          results:
          - description: this stores the status
            name: status
          steps:
          - image: ubuntu
            name: shell-example
            resources: {}
            script: |
              #!/bin/bash
              echo "This is first task"
    test-when-expression-pipeline-finally-run-vgvsp-second-ta-8k64m:
      pipelineTaskName: second-task
      status:
        completionTime: "2020-09-10T16:01:17Z"
        conditions:
        - lastTransitionTime: "2020-09-10T16:01:17Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: test-when-expression-pipeline-finally-run-vgvsp-second-ta-52kbv
        startTime: "2020-09-10T16:01:02Z"
        steps:
        - container: step-shell-example
          imageID: docker.io/library/ubuntu@sha256:31dfb10d52ce76c5ca0aa19d10b3e6424b830729e32a89a7c6eee2cda2be67a5
          name: shell-example
          terminated:
            containerID: cri-o://c03527d787b2aa70c8bdbcbd873aea70631619e5313882b3f41d262941a7a56c
            exitCode: 0
            finishedAt: "2020-09-10T16:01:17Z"
            reason: Completed
            startedAt: "2020-09-10T16:01:17Z"
        taskSpec:
          description: This is first task
          results:
          - description: this stores the status
            name: status
          steps:
          - image: ubuntu
            name: shell-example
            resources: {}
            script: |
              #!/bin/bash
              echo "This is second task"
[varadhya@localhost workspace-testing]$ 
```

### LINKED ISSUE #3203: If the WhenExpression of the task is evaluated to false immediately after the pipelinerun is started, the pipelinerun stays in "Running(Started)" state
# Expected Behavior
The Pipelinerun should start and run successfully 

# Actual Behavior
If the WhenExpression of the task is evaluated to false immediately after the pipelinerun is started, the pipelinerun stays in "Running(Started)" state for a pipeline which has sequential tasks (tasks which run after another)

# Steps to Reproduce the Problem

1. Create 2 tasks (task-1, task-2)
2. Create a pipeline in such a way that task-2 runs after the task-1 and add WhenExpression to the task-2
3. Make sure that the WhenExpression of the task-2 always fails and always gets executed immediately when the pipelinerun is started (you can use static input to make sure that the WhenExpression always fails)
4. Start the pipeline and check the status of the pipelinerun

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
  Client Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0+b3b92b2", GitCommit:"b3b92b2", GitTreeState:"clean", BuildDate:"2020-07-15T09:27:21Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.3+8b0a82f", GitCommit:"8b0a82f", GitTreeState:"clean", BuildDate:"2020-07-10T05:34:00Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version:
  ```
  $ tkn version
  Client version: 0.12.0
  Pipeline version: v0.16.0
  Triggers version: unknown
  ```

**Command Execution Logs**:
```
$ kubectl get pipeline pipeline-with-when-expression -o yaml
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  creationTimestamp: "2020-09-11T05:38:05Z"
  generation: 18
  managedFields:
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        .: {}
        f:params: {}
        f:resources: {}
        f:tasks: {}
    manager: Mozilla
    operation: Update
    time: "2020-09-11T08:35:28Z"
  name: pipeline-with-when-expression
  namespace: veeresh-testing
  resourceVersion: "412847"
  selfLink: /apis/tekton.dev/v1beta1/namespaces/veeresh-testing/pipelines/pipeline-with-when-expression
  uid: 3bc57b32-e865-4d57-ac0d-d32ad235898b
spec:
  params:
  - name: status-1
    type: string
  tasks:
  - name: task-1
    taskRef:
      kind: Task
      name: first-task
  - name: task-2
    runAfter:
    - task-1
    taskRef:
      kind: Task
      name: second-task
    when:
    - Input: $(params.status-1)
      Operator: in
      Values:
      - Success
      - success
$ 
$ 
$ tkn p start pipeline-with-when-expression -p status-1=Failure
PipelineRun started: pipeline-with-when-expression-run-6bd55

In order to track the PipelineRun progress run:
tkn pipelinerun logs pipeline-with-when-expression-run-6bd55 -f -n veeresh-testing
$ 
$ kubectl get pipelinerun
NAME                                      SUCCEEDED   REASON    STARTTIME   COMPLETIONTIME
pipeline-with-when-expression-run-6bd55   Unknown     Started   100s        
$ 
$ tkn pr ls
NAME                                      STARTED        DURATION   STATUS
pipeline-with-when-expression-run-6bd55   1 minute ago   ---        Running(Started)
$ 
$ kubectl get pipelinerun pipeline-with-when-expression-run-6bd55 -o yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  creationTimestamp: "2020-09-11T08:44:37Z"
  generateName: pipeline-with-when-expression-run-
  generation: 1
  labels:
    tekton.dev/pipeline: pipeline-with-when-expression
  managedFields:
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .: {}
          f:tekton.dev/pipeline: {}
      f:status:
        .: {}
        f:conditions: {}
        f:pipelineSpec:
          .: {}
          f:params: {}
          f:tasks: {}
        f:startTime: {}
    manager: controller
    operation: Update
    time: "2020-09-11T08:44:37Z"
  - apiVersion: tekton.dev/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
      f:spec:
        .: {}
        f:params: {}
        f:pipelineRef:
          .: {}
          f:name: {}
    manager: tkn
    operation: Update
    time: "2020-09-11T08:44:37Z"
  name: pipeline-with-when-expression-run-6bd55
  namespace: veeresh-testing
  resourceVersion: "414792"
  selfLink: /apis/tekton.dev/v1beta1/namespaces/veeresh-testing/pipelineruns/pipeline-with-when-expression-run-6bd55
  uid: 5baedc5c-78b9-49ec-b05e-64a09c9156f9
spec:
  params:
  - name: status-1
    value: Failure
  pipelineRef:
    name: pipeline-with-when-expression
  timeout: 1h0m0s
status:
  conditions:
  - lastTransitionTime: "2020-09-11T08:44:37Z"
    reason: Started
    status: Unknown
    type: Succeeded
  pipelineSpec:
    params:
    - name: status-1
      type: string
    tasks:
    - name: task-1
      taskRef:
        kind: Task
        name: first-task
    - name: task-2
      runAfter:
      - task-1
      taskRef:
        kind: Task
        name: second-task
      when:
      - Input: $(params.status-1)
        Operator: in
        Values:
        - Success
        - success
  startTime: "2020-09-11T08:44:37Z"
$ 
$ 
$ kubectl get pipelinerun
NAME                                      SUCCEEDED   REASON    STARTTIME   COMPLETIONTIME
pipeline-with-when-expression-run-6bd55   Unknown     Started   3m6s        
$ 
$ tkn pr ls
NAME                                      STARTED         DURATION   STATUS
pipeline-with-when-expression-run-6bd55   4 minutes ago   ---        Running(Started)
$ tkn tr ls
No TaskRuns found
$ kubectl get taskrun
No resources found.
```

### PR DESCRIPTION
**Pull request by pritidesai**
# Changes

1. Instead of creating a new empty array and appending resolved when expression to it, creating an array based on the existing one and replacing in place.

2. With when expressions having reference to task results, evaluating and applying task results before checking if it's time to run finally tasks. This change is necessary to make sure all the leaf nodes of DAG have resolved task results in their when expressions to check whether DAG is done or not. This change of order works today since finally tasks does not need to resolve task results yet. Once we add task results support in finally, we will have to make one more call to apply task results after adding finally tasks to the execution queue.

3. Do not evaluate when expressions until all parents are done, wait for parent tasks to be completed before evaluating when expressions. This check will make sure that tasks resulting in when expressions failures, are not declared skipped in advance i.e. before their parents (declared with `runAfter`) are started/done.

Fixes #3196
Fixes #3203

The following pipelines are fixed with these changes:

[when-expression-params-with-finally.yaml](https://gist.github.com/pritidesai/605f45336997d5a9ad37586d6558b1cb)
[when-expression-task-results-with-finally.yaml](https://gist.github.com/pritidesai/5d496730976d30e4f4f523d60aba7d5e)
[when-expression-evaluating-to-false-with-runafter.yaml](https://gist.github.com/pritidesai/401e420ff93746adb0e3147c614c45e9)

/kind bug 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Fix to honor When expressions in presence of finally tasks. When expression parameters were not getting resolved in presence of finally tasks and were causing check to skip always return false. Applied fix to when expression parameters resolution function to make sure parameters are substituted with its respective values.
Fix pipelinerun to apply task results to the leaf nodes of DAG before checking if its time to run finally tasks so that when expressions have resolved task results before the check.
Fix pipelinerun to detect and terminate by only evaluating when expressions when all its parents are visited/done instead of evaluating in advance. 
```

### CODE REVIEW

**Comment by pritidesai** [comment]
@bobcatfish @jerop these changes fix both the issues mentioned above. PTAL 🙏

File: pkg/apis/pipeline/v1beta1/when_types.go [hunk 1]
```diff
@@ -107,9 +107,9 @@ func (wes WhenExpressions) HaveVariables() bool {
 // ReplaceWhenExpressionsVariables interpolates variables, such as Parameters and Results, in
 // the Input and Values. Returns a new instance of WhenExpressions.
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1beta1/when_types.go:108]: `Returns a new instance of WhenExpressions` is not true anymore, is it ?

Comment (bobcatfish) [reply pkg/apis/pipeline/v1beta1/when_types.go:108]: I think we should move toward LESS mutation and side effects in these functions vs more

**Comment by bobcatfish** [comment]
Thanks for getting a fix in for this - @jerop and i were working on a fix as well, i think @jerop is checking to see if this fixes the examples we'd been working with as well. if it does we can get this in and do the 0.16.2 release and after that i think its important that we step back and try to make the code around this less complex so its easier to work with in the future

**Comment by jerop** [comment]
@pritidesai thank you! 

tested with the cases @bobcatfish and I were looking at and they pass

can we please modify the `pipelinerun-with-when-expressions` example in this pr to contain `finally` tasks and `runAfter` tasks, like this: https://gist.github.com/jerop/f9a6d46140c30aad6ca9006eedc02874

File: pkg/apis/pipeline/v1beta1/when_types.go (hunk 1 above)
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/when_types.go:108]: right, I will update the comment, there is no need to create new instances here. The goal is to get the params/task results resolved which can be effectively done in place.

Comment (bobcatfish) [reply pkg/apis/pipeline/v1beta1/when_types.go]: probably better we don't act on this in this PR but I'm advocating for the opposite.

while I don't think we should try to program in a purely functional way, I think limiting side effects and having more "pure" functions (https://dev.to/jamesrweb/principles-of-functional-programming-4b7c) leads to code that is easier to reason about and maintain over time

"functional core, imperative shell" is another approach i like: basically the idea is to limit mutations and other side effects in the code so you know exactly where they are happening.

in this specific example, its actually not clear to me why making "ReplaceWhenExpressionsVariables" mutate state instead of returning a new object is fixing anything - in the places where this function is being called (https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L81 https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L100) the returned value is being used - the core problem is probably that it isn't being used consistently, e.g. in ApplyReplacements `p` is being returned although `p` is not being directly mutated:

https://github.com/tektoncd/pipeline/blob/c12b84311e6fa5b7bc9c3d005c705c9944909be3/pkg/reconciler/pipelinerun/resources/apply.go#L87-L104

I think the inconsistent behavior we are seeing is because of the append semantics (`tasks := append(p.Tasks, p.Finally...)`) that we've debated previously - but i think the answer is not MORE mutation, i think it is relying on it LESS

anyway long story short, although we dont NEED to create new instances, I think this is a cleaner approach in the long run, but let's follow up on that in subsequent PRs
